src/ui/src/components/workflows/WorkflowsWorkflow.vue

```diff
@@ -209,55 +235,69 @@ function handleNodeDrag(ev: DragEvent, componentId: Component["id"]) {
 	if (x < 0 || y < 0) return;
 
 	const component = wf.getComponentById(componentId);
-	component.x = x - clickOffset.x;
-	component.y = y - clickOffset.y;
-	refreshArrows();
+
+	const newX = x - clickOffset.x;
+	const newY = y - clickOffset.y;
+
+	if (component.x == newX && component.y == newY) return;
+
+	component.x = newX;
+	component.y = newY;
+
+	setTimeout(() => {
+		// Debouncing
+		if (component.x !== newX) return;
+		if (component.y !== newY) return;
+		changeCoordinates(componentId, newX, newY);
+	}, 200);
 }
 
 function handleNodeDragend(ev: DragEvent) {
 	ev.preventDefault();
 }
 
-function handleNodeClick(ev: MouseEvent, componentId: Component["id"]) {
+async function handleNodeClick(ev: MouseEvent, componentId: Component["id"]) {
 	if (!activeNodeOut.value) return;
 	if (activeNodeOut.value.fromComponentId == componentId) return;
 
-	const component = wf.getComponentById(activeNodeOut.value.fromComponentId);
-
-	component.outs = [
-		...(component.outs ?? []),
-		{
-			toNodeId: componentId,
-			outId: activeNodeOut.value.outId,
-		},
-	];
+	addOut(activeNodeOut.value.fromComponentId, {
+		toNodeId: componentId,
+		outId: activeNodeOut.value.outId,
+	});
 
 	activeNodeOut.value = null;
-	refreshArrows();
 }
 
-function createNode(type: string, x: number, y: number) {
-	const componentId = createAndInsertComponent(type, workflowComponentId);
-	const component = wf.getComponentById(componentId);
-	component.x = x;
-	component.y = y;
+async function createNode(type: string, x: number, y: number) {
+	createAndInsertComponent(type, workflowComponentId, undefined, { x, y });
 }
 
-watch(children, async (postChildren, preChildren) => {
-	// Remove references when a node is deleted
+watch(
+	children,
+	async (postChildren, preChildren) => {
+		// Refresh arrows
 
-	const preIds = preChildren.map((c) => c.id);
-	const postIds = postChildren.map((c) => c.id);
-	const removedIds = preIds.filter((cId) => !postIds.includes(cId));
-	removedIds.forEach((removedId) => {
-		postChildren.forEach((c) => {
-			if (!c.outs || c.outs.length == 0) return;
-			c.outs = c.outs.filter((out) => out.toNodeId !== removedId);
+		await nextTick();
+		refreshArrows();
+
+		// Remove references when a node is deleted
+
+		const preIds = preChildren.map((c) => c.id);
+		const postIds = postChildren.map((c) => c.id);
+		const removedIds = preIds.filter((cId) => !postIds.includes(cId));
+		removedIds.forEach((removedId) => {
+			postChildren.forEach((c) => {
+				if (!c.outs || c.outs.length == 0) return;
+				c.outs = c.outs.filter((out) => out.toNodeId !== removedId);
+			});
```

Comment:
I think you can avoid the `O(n²)` and I guess it makes sens to use `Set`. Maybe like this (not tested)
```suggestion
    const removedIds = new Set(postChildren.map((c) => c.id));
    preChildren.forEach(c => removedIds.delete(c.id));

    postChildren.forEach((c) => {
      c.outs = c.outs?.filter((out) => !removedIds.has(out.toNodeId));
    });
```
There is also the new method [`Set.difference`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set/difference) but still not widely available ([caniuse](https://caniuse.com/?search=set%20difference))

Reply:
I implemented the suggestion but it wasn't working, I asked Palmyra to break it down.

-------

 The issue with the second alternative triggering an infinite recursion likely stems from how the `removedIds` set is being manipulated and used. Let's break down the differences between the two alternatives to understand why the second one causes an infinite loop.

### First Alternative

In the first alternative, the `removedIds` array is calculated by comparing the `preIds` and `postIds` arrays. This ensures that only the IDs that are present in `preChildren` but not in `postChildren` are considered as removed.

```javascript
const preIds = preChildren.map((c) => c.id);
const postIds = postChildren.map((c) => c.id);
const removedIds = preIds.filter((cId) => !postIds.includes(cId));
```

Then, it iterates over `postChildren` and filters out any `outs` that reference the removed IDs.

```javascript
removedIds.forEach((removedId) => {
    postChildren.forEach((c) => {
        if (!c.outs || c.outs.length == 0) return;
        c.outs = c.outs.filter((out) => out.toNodeId !== removedId);
    });
});
```

### Second Alternative

In the second alternative, the `removedIds` set is initially populated with the IDs from `postChildren`, and then the IDs from `preChildren` are removed from this set.

```javascript
const removedIds = new Set(postChildren.map((c) => c.id));
preChildren.forEach((c) => removedIds.delete(c.id));
```

However, this logic is flawed because it ends up with a set of IDs that are present in `postChildren` but not in `preChildren`, which is the opposite of what you want. You want the IDs that are present in `preChildren` but not in `postChildren`.

Additionally, the filtering logic is applied directly to `postChildren`, which might cause the watcher to trigger again if the `outs` array is modified.

```javascript
postChildren.forEach((c) => {
    c.outs = c.outs?.filter((out) => !removedIds.has(out.toNodeId));
});
```

### Why Infinite Recursion?

The infinite recursion occurs because the watcher is set to `deep: true`, which means it watches for changes in nested properties. When you modify `c.outs` within the watcher, it triggers the watcher again, leading to an infinite loop.

### Corrected Second Alternative

To fix the second alternative, you should correctly calculate the `removedIds` set and avoid modifying `postChildren` directly within the watcher. Here's a corrected version:

```javascript
watch(
    children,
    async (postChildren, preChildren) => {
        // Remove references when a node is deleted

        const preIds = new Set(preChildren.map((c) => c.id));
        const postIds = new Set(postChildren.map((c) => c.id));
        const removedIds = new Set([...preIds].filter((cId) => !postIds.has(cId)));

        postChildren.forEach((c) => {
            if (!c.outs || c.outs.length === 0) return;
            c.outs = c.outs.filter((out) => !removedIds.has(out.toNodeId));
        });

        // Refresh arrows

        await nextTick();
        refreshArrows();

        if (removedIds.size > 0) {
            wf.sendComponentUpdate();
        }
    },
    { deep: true },
);
```

In this corrected version, `removedIds` is calculated correctly, and the filtering logic is applied without causing the watcher to trigger again. This should prevent the infinite recursion.